Document update for MaxParallelImagePulls (#39732)
kubernetes/website: content/en/docs/concepts/containers/images.md

Kubernetes raises the delay between each attempt until it reaches a compiled-in limit,
which is 300 seconds (5 minutes).

## Serial and parallel image pulls

> By default, kubelet pulls images serially. In other words, kubelet will send only
> one image pull request to the image service at a time. Other image pull requests
> have to wait until the one being processed is complete.

Reviewer: Is this right:
```suggestion
By default, kubelet pulls images serially. In other words, kubelet will send only
one image pull request to the image service at a time. Other image pull requests
have to wait until the one being processed is complete.

If two or more Pods on a node use the same image, the kubelet only makes one image
pull request on behalf of the local Pods. The kubelet on a given node does not coordinate
image pulls with other nodes in your cluster: different nodes can always start image pulls from
the same registry in parallel.

```

Author: > For a given container image, the kubelet will send only one image pull request at a time.

This is not only for one given container image. Kubelet sends only one pull request across all pods on the node.

If you would like to enable parallel image pulls, you can set the field
`serializeImagePulls` to false in kubelet configuration. With `serializeImagePulls`
set to false, image pull requests will be sent to the image service immediately,

> and multiple images will be pulled at the same time.

Reviewer 2: please note the difference of parallel image pulls from different pods. Images from the same pods are never pulled in parallel

### Maximum parallel image pulls
{{< feature-state for_k8s_version="v1.27" state="Alpha" >}}
By default, when `serializeImagePulls` is set to false, kubelet does not limit the
maximum number of images being pulled at the same time. If you would like to limit

> the number of parallel image pulls, you can set the field `maxParallelImagePulls`

Reviewer 2: This bare documentation it technically OK and correct. So lgtm at will.

What I would suggest as an improvement is to explain why one will want to limit this number

Author: Added a short paragraph on why a user might want to limit this number

in kubelet configuration. With `maxParallelImagePulls` set to `N`, only `N` images
can be pulled at the same time, and any image pull beyond `N` will have to wait

> until at least one ongoing image pull is complete.

Reviewer 2: I'd suggest to mention that the same image pull request will stay in line waiting for it's turn. There is no deduplication on kubelet side

Reviewer: I amended the suggestion in https://github.com/kubernetes/website/pull/39732#discussion_r1139172957 to account for this.

> 

Reviewer: ```suggestion

The kubelet never pulls multiple images in parallel on behalf of one Pod. If you have two
multi-container Pods that are replicas (and therefore use the same container images), the kubelet
may pull the different images in parallel on behalf of the two different Pods.

```

The field `maxParallelImagePulls` can only be positive numbers. The field
`serializeImagePulls` must be set to false, if `maxParallelImagePulls` is equal

> or larger than 2.

Author: Validation on kubelet config will fail, and kubelet will not start. I can mention this here.

> or larger than 2.

Author: Adding "Kubelet will fail to start with invalid `maxParallelImagePulls` settings."

> 

Reviewer 2: can we add a note on how container runtime needs to be configured to handle parallel image pull? Maybe just mention - check with the runtime configuraiton to ensure it can handle this and configured properly.

@sftim would it make sense to point to specific Containerd settings that would control it?

Author: I don't think there is much configuration needed on the container runtime. But I will just mention that make sure container runtime can handle parallel image pulls.

Kubernetes raises the delay between each attempt until it reaches a compiled-in limit,
which is 300 seconds (5 minutes).

## Serial and parallel image pulls

By default, kubelet pulls images serially. In other words, kubelet will send only
one image pull request to the image service at a time. Other image pull requests
have to wait until the one being processed is complete.

Nodes make image pull decisions in isolation. Even when you use serialized image pulls,

> two different nodes can pull the same image in parallel.

Reviewer 2: In my other comment I wanted to highlight that when Pod has Init contianer and regular container, image download for those will not be parallelized. Those will be sequencial. Can you please add this note?

If you would like to enable parallel image pulls, you can set the field
`serializeImagePulls` to false in kubelet configuration. With `serializeImagePulls`
set to false, image pull requests will be sent to the image service immediately,
and multiple images will be pulled at the same time.

> When enabling parallel image pulls, please make sure the container runtime can handle

Reviewer 2: ```suggestion
When enabling parallel image pulls, please make sure the image service of your container runtime can handle
```

which is 300 seconds (5 minutes).

## Serial and parallel image pulls

By default, kubelet pulls images serially. In other words, kubelet sends only
one image pull request to the image service at a time. Other image pull requests
have to wait until the one being processed is complete.

Nodes make image pull decisions in isolation. Even when you use serialized image
pulls, two different nodes can pull the same image in parallel.

If you would like to enable parallel image pulls, you can set the field
`serializeImagePulls` to false in the [kubelet configuration](/docs/reference/config-api/kubelet-config.v1beta1/).
With `serializeImagePulls` set to false, image pull requests will be sent to the image service immediately,
and multiple images will be pulled at the same time.

When enabling parallel image pulls, please make sure the image service of your
container runtime can handle parallel image pulls.

The kubelet never pulls multiple images in parallel on behalf of one Pod. For example,
if you have a Pod that has an init contianer and an application container, the image
pulls for the two containers will not be parallelized. However, if you have two

> Pods that use differrent images, the kubelet pulls the images in parallel on

Reviewer 3: line 180: differrent/different
line 178: contianer/container

behalf of the two different Pods, when parallel image pulls is enabled.

### Maximum parallel image pulls

{{< feature-state for_k8s_version="v1.27" state="alpha" >}}

When `serializeImagePulls` is set to false, the kubelet defaults to no limit on the
maximum number of images being pulled at the same time. If you would like to
limit the number of parallel image pulls, you can set the field `maxParallelImagePulls`
in kubelet configuration. With `maxParallelImagePulls` set to _n_, only _n_ images
can be pulled at the same time, and any image pull beyond _n_ will have to wait
until at least one ongoing image pull is complete.

Limiting the number parallel image pulls would prevent image pulling from consuming

> too much network bandwith or disk I/O, when parallel image pulling is enabled.

Reviewer 3: typo: bandwidth

Kubernetes raises the delay between each attempt until it reaches a compiled-in limit,
which is 300 seconds (5 minutes).

## Serial and parallel image pulls

> By default, kubelet pulls images serially. In other words, kubelet will send
> only

Reviewer 4: ```suggestion
By default, kubelet pulls images serially. In other words, kubelet sends only
```

> one image pull request to the image service at a time. Other image pull requests

Reviewer 4: What do you mean by 'image' service? Are you talking about the component inside kubelet?
Users may get confused because they don't know the internals and they are not interested in them.

```suggestion
one image pull request to the image registry at a time. Other image pull requests
```

Author: Image service is the service that handles all image requests from kubelet. Similar to container runtime, it is something configurable on kubelet. See `--image-service-endpoint` on https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

have to wait until the one being processed is complete.

Nodes make image pull decisions in isolation. Even when you use serialized image

> pulls,
> two different nodes can pull the same image in parallel.

Reviewer 4: ```suggestion
pulls, two different nodes can pull the same image in parallel.
```

If you would like to enable parallel image pulls, you can set the field

> `serializeImagePulls` to false in kubelet configuration. With
> `serializeImagePulls`

Reviewer 4: ```suggestion
`serializeImagePulls` to false in the [kubelet configuration](/docs/reference/config-api/kubelet-config.v1beta1/).
With `serializeImagePulls`set to false, image pull requests will be sent to the image service immediately,
and multiple images will be pulled at the same time.
```

> set to false, image pull requests will be sent to the image service immediately,
> and multiple images will be pulled at the same time.

Reviewer 4: ```suggestion
```

When enabling parallel image pulls, please make sure the image service of your

> container runtime can handle parallel image pulls.

Reviewer 4: Again, the 'image' service of your container runtime ... what is that?

Author: same as above.

> The kubelet never pulls multiple images in parallel on behalf of one Pod. For
> example,

Reviewer 4: ```suggestion
The kubelet never pulls multiple images in parallel on behalf of one Pod. For example,
```

> if you have a Pod that has an init contianer and a regular container, the image

Reviewer 4: ```suggestion
if you have a Pod that has an init container and an application container, the image
```

pulls for the two containers will not be parallelized. However, if you have two

> Pods that use differrent images, the kubelet will pull the images in parallel on

Reviewer 4: ```suggestion
Pods that use differrent images, the kubelet pulls the images in parallel on
```

behalf of the two different Pods, when parallel image pulls is enabled.

> ### Maximum parallel image pulls
> {{< feature-state for_k8s_version="v1.27" state="alpha" >}}
> When `serializeImagePulls` is set to false, the kubelet defaults to no limit on
> the

Reviewer 4: ```suggestion
### Maximum parallel image pulls

{{< feature-state for_k8s_version="v1.27" state="alpha" >}}

When `serializeImagePulls` is set to false, the kubelet defaults to no limit on the
```

maximum number of images being pulled at the same time. If you would like to

> limit
> the number of parallel image pulls, you can set the field
> `maxParallelImagePulls`
> in kubelet configuration. With `maxParallelImagePulls` set to _n_, only _n_
> images
> can be pulled at the same time, and any image pull beyond _n_ will have to wait

Reviewer 4: ```suggestion
limit the number of parallel image pulls, you can set the field `maxParallelImagePulls`
in kubelet configuration. With `maxParallelImagePulls` set to _n_, only _n_ images
can be pulled at the same time, and any image pull beyond _n_ will have to wait
```

until at least one ongoing image pull is complete.

> Limiting the number parallel image pulls would prevent image pulling from
> consuming

Reviewer 4: ```suggestion
Limiting the number parallel image pulls would prevent image pulling from consuming
```

too much network bandwith or disk I/O, when parallel image pulling is enabled.

You can set `maxParallelImagePulls` to a positive number that is greater than or

> equal
> to 1. If you set `maxParallelImagePulls` to be greater than or equal to 2, you

Reviewer 4: ```suggestion
equal to 1. If you set `maxParallelImagePulls` to be greater than or equal to 2, you
```

> must set the
> `serializeImagePulls` to false. The kubelet will fail to start with invalid

Reviewer 4: ```suggestion
must set the `serializeImagePulls` to false. The kubelet will fail to start with invalid
```

> `maxParallelImagePulls`
> settings.

Reviewer 4: ```suggestion
`maxParallelImagePulls` settings.
```
